// hci-hospindex-api/webpack.config.js
const webpack = require('webpack');
const path = require('path');
const PROD = JSON.parse(process.env.PROD || '0');


module.exports = {
    entry: __dirname + '/src/index.ts',
    output: {
        path: __dirname + '/dist',
        filename: 'index.js',
        library: 'hci-hospindex-api',
        libraryTarget: 'umd',
        umdNamedDefine: true
    },
    devtool: 'source-map', // for debugging purposes
    resolve: {
        extensions: [ '.ts', '.js' ],
        alias: {
            '@hci-hospindex-api': path.resolve('./src')
        }
    },
    module: {
        loaders: [
            { test: /\.tsx?$/, loader: 'ts-loader?logLevel=warn', exclude: /node_modules/}
        ]
    },
    resolve: {
        extensions: [".tsx", ".ts", ".js"]
    },
    plugins: PROD ? [
        new webpack.optimize.UglifyJsPlugin({
            minimize: true,
            sourceMap: true,
            output: {
                comments: false
            },
            compress: { warnings: false }
        })
    ] : []
};
